### PR TITLE
Device: Improvements to GPU CDI mode & Ubuntu Core validation

### PR DESCRIPTION
**Pull request by mihalicyn**
https://github.com/canonical/lxd/pull/14347#discussion_r1816694946

### CODE REVIEW

File: lxd/device/gpu_physical.go [hunk 1]
```diff
@@ -75,6 +75,12 @@ func (d *gpuPhysical) validateConfig(instConf instance.ConfigReader) error {
 				return fmt.Errorf(`Cannot use %q when "id" is set`, field)
 			}
 		}
+
+		// If user requests CDI in conjunction with nvidia.runtime=true we should forbid that.
+		_, err := cdi.ToCDI(d.config["id"])
+		if err == nil && shared.IsTrue(instConf.ExpandedConfig()["nvidia.runtime"]) {
```
Comment (tomponline) [inline lxd/device/gpu_physical.go:81]: we could make this a bit more efficient by testing `shared.IsTrue(instConf.ExpandedConfig()["nvidia.runtime"]) ` first and then checking `cdi.ToCDI` (as that does more parsing).

Comment (mihalicyn) [reply lxd/device/gpu_physical.go:81]: sure, I'll rework that one. Thanks!

Comment (gabrielmougard) [reply lxd/device/gpu_physical.go:81]: Yes I agree with @tomponline :)

Comment (mihalicyn) [reply lxd/device/gpu_physical.go]: >We should also check if if d.config["id"] != "" and not check either if id not specified, as that indicates CDI mode cannot be in use.

We are inside `if d.config["id"] != ""` block already ;-)

File: lxd/device/gpu_physical.go [hunk 2]
```diff
@@ -75,6 +75,14 @@ func (d *gpuPhysical) validateConfig(instConf instance.ConfigReader) error {
 				return fmt.Errorf(`Cannot use %q when "id" is set`, field)
 			}
 		}
+
+		// If user requests CDI in conjunction with nvidia.runtime=true we should forbid that.
+		if shared.IsTrue(instConf.ExpandedConfig()["nvidia.runtime"]) {
```
[the remark this answers is not in the document]
Comment (mihalicyn) [reply lxd/device/gpu_physical.go:80]: we are inside  `if d.config["id"] != "" {` block ;-)

[the remark this answers is not in the document]
Comment (tomponline) [reply lxd/device/gpu_physical.go:80]: oh yes indeed, sorry

**Review by tomponline — APPROVED** [review]

**Comment by tomponline** [comment]
Thanks!

**Pull request merged**